pkg/services/grpcserver/datasource.go

```diff
@@ -0,0 +1,155 @@
+package grpcserver
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana-plugin-sdk-go/genproto/pluginv2"
+	"github.com/grafana/grafana/pkg/components/simplejson"
+	"github.com/grafana/grafana/pkg/expr"
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/plugins"
+	"github.com/grafana/grafana/pkg/plugins/adapters"
+	"github.com/grafana/grafana/pkg/services/datasources"
+	"github.com/grafana/grafana/pkg/services/user"
+	"github.com/grafana/grafana/pkg/setting"
+	"github.com/grafana/grafana/pkg/tsdb/grafanads"
+)
+
+type DatasourceService struct {
+	cfg    *setting.Cfg
+	server *DatasourceServer
+}
+
+func ProvideDatasourceService(cfg *setting.Cfg, grpcServerProvider Provider, pluginsClient plugins.Client, dataSourceCache datasources.CacheService, dataSourceService datasources.DataSourceService) (*DatasourceService, error) {
+	server := &DatasourceServer{
+		logger:            log.New("grpc-server-datasource"),
+		pluginClient:      pluginsClient,
+		dataSourceCache:   dataSourceCache,
+		dataSourceService: dataSourceService,
+	}
+	pluginv2.RegisterDataServer(grpcServerProvider.GetServer(), server)
+	return &DatasourceService{
+		cfg:    cfg,
+		server: server,
+	}, nil
+}
+
+type DatasourceServer struct {
+	logger            log.Logger
+	pluginClient      plugins.Client
+	dataSourceCache   datasources.CacheService
+	dataSourceService datasources.DataSourceService
+}
+
+// QueryData handles multiple queries and returns multiple responses.
+func (s *DatasourceServer) QueryData(ctx context.Context, in *pluginv2.QueryDataRequest) (*pluginv2.QueryDataResponse, error) {
+	req := backend.FromProto().QueryDataRequest(in)
+	user := adapters.SignedInUserFromBackendUser(req.PluginContext.User)
```

Comment:
For auth...  can we get the user from initial token? (API_KEY)

Long term we can ideally get this from context (and properties sent in the request)

Reply:
Right now it is better to use/force the admin API key than to let requests delegate to a different user

Reply:
> Long term we can ideally get this from context (and properties sent in the request)

@ryantxu my thinking was that using the service account token for authN and the plugin context user for authZ allows the query request to respect the permission of the parent request. it seems like querying through the gRPC server using the service account token for authZ could allow a user to access additional datasources that they wouldn't have access to via the HTTP API. if the plugin wanted to use the service account user, it could always set that user in the plugin context. is that kind of what you were thinking?

Reply:
@ryantxu bubbling up the conversation from the daily before marking this as resolved. it sounds like you are good with this approach for now since it's behind a feature flag and requires an admin service account token, correct?

Reply:
I don't see where it "requires an admin service account token" -- but I could be missing something

Reply:
Ahhh -- I see, if you do not implement custom auth, then it requires an admin token 👍 

I think this is fine until we have a more robust token/delegation system in place